README.md
# Jupyter Graffiti

Create interactive screencasts inside Jupyter Notebook that anybody can play back.

![intro_movie](./images/tutorial/graffitiMoviePlay1.gif)

Ever wanted to offer someone a hands-on "live" demo in Jupyter
Notebook? Using __Graffiti__ you can add floating tips along with
optional recorded movies anywhere in your Notebooks. Demonstrate
whatever you want, highlight what's important-- all narrated by you!
Since the movie plays back in and _on_ the Notebook, the viewer can
pause any time, dive into your code and try things out exactly as you
had them when you were recording, and then resume play.

## Try Graffiti Out! Try These _Live_ Demos:

* [![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/willkessler/jupytergraffiti/master?filepath=samples%2Fgeneral%2FGraffiti%20Basic%20Demo.ipynb) <a href="https://mybinder.org/v2/gh/willkessler/jupytergraffiti/master?filepath=samples%2Fgeneral%2FGraffiti%20Basic%20Demo.ipynb">General Demo</a>
* [![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/willkessler/jupytergraffiti/master?filepath=samples%2Fudacity%2Fc%2B%2B_nanodegree%2FFor%20Loops.ipynb) <a href="https://mybinder.org/v2/gh/willkessler/jupytergraffiti/master?filepath=samples%2Fudacity%2Fc%2B%2B_nanodegree%2FFor%20Loops.ipynb">Udacity C++ Nanodegree Example: "For Loops"</a>
* [![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/willkessler/jupytergraffiti/master?filepath=samples%2Fudacity%2Fc%2B%2B_nanodegree%2FPointers.ipynb) <a href="https://mybinder.org/v2/gh/willkessler/jupytergraffiti/master?filepath=samples%2Fudacity%2Fc%2B%2B_nanodegree%2FPointers.ipynb">Udacity C++ Nanodegree Example: "Pointers"</a>
* [![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/willkessler/jupytergraffiti/master?filepath=samples%2Fudacity%2Fdata_structures_nanodegree%2FImplement%20a%20stack%20using%20an%20array.ipynb) <a href="https://mybinder.org/v2/gh/willkessler/jupytergraffiti/master?filepath=samples%2Fudacity%2Fdata_structures_nanodegree%2FImplement%20a%20stack%20using%20an%20array.ipynb">Udacity Data Structures and Algorithms Nanodegree Example: "Implementing a Stack using an Array"</a>

_Please wait about 30 seconds for the demonstration to spin up at mybinder.org. (Thanks to the folks at MyBinder for this awesome service)._

## Or, Watch Regular Recorded Video Demos:

* <a target="_blank" href="https://youtu.be/3hGD0PiPuhI">Gentle Introduction To Graffiti</a> (Start here)
* <a target="_blank" href="https://www.youtube.com/watch?v=8gDmr5MiIVA">How to make Graffiti movies.</a>
* <a target="_blank" href="https://www.youtube.com/watch?v=Mwh5rCRXar4">How to make a Graffiti tooltip in code</a>
* <a target="_blank" href="https://www.youtube.com/watch?v=PUD-Juu0ZQE&t=9s">Using Graffiti extras like in-line terminals and auto-saving code cells.</a>
* <a target="_blank" href="https://www.youtube.com/watch?v=hjU0gmis078">Adding a recording to a Show/Hide solution button.</a>
* <a target="_blank" href="https://www.youtube.com/watch?v=3Zu2Wu44N_I&t=1s">How Graffiti in-line shells are recorded into movies.</a>
* <a target="_blank" href="https://www.youtube.com/watch?v=vBNHcqJCW-Q">Accessing and using the Graffiti API</a>

## What exactly can you record in the Graffiti?

* Recorded audio (e.g. voice narration), captured with your laptop's microphone while making your recording
* Mouse movement and scrolling in the Notebook
* Selecting and editing inside code cells
* The output of any code cell executions
* Inlined terminals (shells) whose activities you can also record.
* You can also draw and highlight over sections you think are important, or create handwritten notes.
* Support for the <a href="https://github.com/QuantStack/xeus-cling">C++ kernel</a> and the <a href="https://github.com/IRkernel/IRkernel">R kernel</a> are included in addition to the Python kernel.

## Learn More On Using Graffiti:

* <a target="_blank" href="https://drive.google.com/file/d/1eEga9NJ4ak8RZh28co4CZRck-4TrdjsE/view">How to make Graffiti movies</a>
* <a target="_blank" href="https://drive.google.com/file/d/1VP5U8xOYsS9rRM4-nc7IjKiLkubBu5Q3/view">How to make a Graffiti tooltip in code</a>
* <a target="_blank" href="https://drive.google.com/file/d/1y3bXMF5nvUYP9E67SATotRw8m_TRjg4O/view">Accessing and using the Graffiti API</a>
* <a target="_blank" href="https://drive.google.com/file/d/1i6QuWlHfLNu62d0l1CIteiP-1yNqaHuY/view">Using Graffiti extras like in-line terminals and auto-saving code cells</a>
* <a target="_blank" href="https://drive.google.com/file/d/1N2gyzCN14b1vTwZBxNjsgNhZklku-qq4/view">Adding a recording to a Show/Hide solution button</a>
* <a target="_blank" href="https://drive.google.com/file/d/1ROb1dCp-w4Js0P77-YcqRWHf0Y0hh0c8/view">How Graffiti in-line shells are recorded into movies</a>

You can also visit the <a href="user_manual/UserManual.ipynb">User Manual<a> for more detailed instructions on how to use Graffiti.

## Advantages of Graffiti Over Traditional Screencasts

* You can save any number of Graffitis in a Notebook.
* You don't need any special software or hardware (other than this library and Chrome/Firefox) to create and save Graffitis.
* Viewers can pause recorded playback any time, scrub forward and
backward, and interact with the Notebook during playback at any
point. No need to watch a whole recorded screencast first, and then
switch context to a Notebook; students can explore right along with
you in the same environment you recorded in. When you pause playback,
you're still in a live Notebook, so you can play around.
* Jupyter Graffiti is easy to set up: either use the Python library
  or build the Docker image with the included Jupyter extension.  (At
  Udacity, Jupyter Notebook Workspaces use the extension. See below
  how to install it). Or, you can skip installation entirely (see below).
* All data, including audio, is stored in compressed plain text in a
  directory separate from your Notebook files, for easy portability and
  storage in any version control system e.g. git/github.
* Unlike streamed video, you don't need a video server or hosted
  YouTube videos, and you can watch the videos even without an
  internet connection or over narrow bandwidth, because the files are
  very small.

## Using Graffiti with ZERO installation

You can skip installation entirely if you want, by using the Binder
demo link above. Just upload whatever Notebook you want to add
Graffiti to the demo Jupyter Notebook server, activate Graffiti
(see below), make some recordings, and then download the
`jupytergraffiti_data` folder along with your Graffiti-ized Notebook
from the binder Notebook server. To make this easier, we've installed
`nbzip` into the Binder demo server. Just go to the server's tree,
visit the `jupytergraffiti_data` folder and click the folder download
link. For more details on how this works you can refer to <a
href="https://github.com/data-8/nbzip">`nbzip`'s</a> documentation. 

You can then commit the `jupytergraffiti_data` folder and your
Notebook to your own github repo and set up a link to it on Binder to
share your Graffiti without requiring that the recipient have Graffiti
installed.

## Graffiti Software Installation-- Putting Graffiti onto your own System(s)

Uploading and downloading to/from Binder isn't optimal of course; it's probably better to get Graffiti going on your own computers.

There are four ways to install Jupyter Graffiti: using pip or conda, using a Python library, using a Docker image, or installing a plugin into your Jupyter Notebook server.

### Installation Option #1: Use `pip` or `conda` (Easiest Option)

Just do:

 `pip install -c willkessler jupytergraffiti`

or

`conda install -c willkessler jupytergraffiti`

This assumes you have `pip` or `pip3`, or `conda` as well as Jupyter Notebook, already installed.


### Installation Option #2: Run Jupyter Notebook with a Docker Image Containing Graffiti (Slightly More Complex Option)

Make sure to **[install Docker](https://docs.docker.com/install)** on your system first (unless you've already installed it).

Then enter this command in a terminal on your computer:

```
./jupytergraffiti/build_and_run.sh
```

This will build and start up a Docker container running the Jupyter
Server and the Jupyter Graffiti extension, with the container's home
directory being mounted where your Jupyter Notebook(s) are located,
and serving Notebooks over port 8888.

The advantage of using the Docker container is that *Jupyter Graffiti*
is always loaded automatically, so you don't have to execute `import
jupytergraffiti` in the Notebook just to play back *Graffitis* (but
you *will* need to run it to access the _Graffiti_ API).

using the Docker container also ensures you're running a recent
version of Jupyter Notebook.

Take a look at the output of the Jupyter Server running in the
container. It has the secret key you need to be able to surf to the Dockerized
Jupyter server. The output will look something like this (but note
that the Jupyter Server login token will change every time you run `build_and_run.sh`):

```
Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://(b13ba2b482e9 or 127.0.0.1):8888/?token=e58a08f167881500e207ff9be05ad57ffe00e3457e54017c
```

What this is telling you is to surf to
`http://localhost:8888?token=e58a08f167881500e207ff9be05ad57ffe00e3457e54017c`
to access the Jupyter Server with the _Graffiti_ extension installed and
running.

The Docker container will serve content out of port 8888. If you already have
something (e.g. another Jupyter server) running on this port, pass a different port to
`build_and_run.sh` like so:

```
./jupytergraffiti/build_and_run.sh 8889

```

**Note**: if you specify a different port, the Jupyter Server output
containing the secret key will still show port 8888, because
internally it still uses port 8888; via Docker port mappings, we
have remapped 8888 to the port you specify. In the example above,
therefore, you would need to access the server at:
`http://localhost:8889?token=e58a08f167881500e207ff9be05ad57ffe00e3457e54017c`.


### Installation Option #3: Use the Python Library (Slightly More Involved Option, but does not add code to your system)

*Note*: Before using this method, you may need to Trust your
Notebook. This is because Jupyter Graffiti is mostly written in Javascript,
and by default, if the Notebook you're adding _Graffitis_ to was not
created by you, Jupyter Notebook will not "Trust" it and will not run
externally loaded javascript code, for security reasons.  To Trust a
Notebook, click `File...Trust Notebook`*before*  running the import
command below.

1. `git clone` this repo in the same directory where you keep the Notebook(s) you want to add _Graffiti_ to.
1. Add and run the following command in a cell in the Notebook you want to start adding _Graffiti_ to:

```
import jupytergraffiti
```

If everything works, you will see a button labelled "Activate Graffiti" in your menu bar.

If you don't see this button appearing, use `Kernel... Restart and Clear Output` first, then try running ```import jupytergraffiti``` again.

![kernel_restart](./images/tutorial/kernel_restart.png)

Once you see this message, you can "Activate Graffiti" on a Notebook
to begin creating Graffiti. The <a
href="user_manual/UserManual.ipynb">User Manual<a> has many details on
how to create Graffiti.

**Special Note** : if you are adding *Graffitis* to Notebooks that do not reside in the same folder where you cloned this repo, then you must :

1. Create a `jupytergraffiti_data` directory in the folder where you cloned this repo (`mkdir jupytergraffiti_data`).
1. Create symbolic links from the directory where you Notebook resides
   to both the `jupytergraffiti` folder in this repo, and to the
   `jupytergraffiti_data` folder alongside where you cloned this repo.

### Installation Option #4: Install the Graffiti Extension in Your Own Jupyter Server (Most Complex Option)

This will permanently install the extension in your computer's
installation of the Jupyter Server (although you can always disable
it if you want to). This means the extension will always be available
whenever you start up your Jupyter server. To install the extension:

```
cd jupytergraffiti
jupyter nbextension install graffiti_extension --symlink --user
jupyter nbextension enable graffiti_extension/main --user
cd ..
```

You may need to restart your Jupyter server to get the extension to load, although usually that's not required.

#### Disabling the Graffiti extension in your Jupyter Server

If you need to disable the _Graffiti_ plugin for some reason, you can easily do it.

To disable the plugin:

```
cd jupytergraffiti
jupyter nbextension disable graffiti_extension/main --user
cd ..
```

Then restart your Jupyter server.

## Using the Jupyter Graffiti Python API

When you `import jupytergraffiti` you get immediate access to
functions you can use to control Jupyter Graffiti from Python. Some
of these are utility functions, and others can be used to control
recordings playback.  To use them, simply run the Python functions in your
Notebook's cells (note: the API only works with the Python kernels at this time).

#### Playing Back Graffiti Recordings

You can also play any Graffiti recording back using Python
code. This could be valuable, for instance, after a student has failed
several times to make some code work; you could watch for this
situation in your testing code, and offer to play a hint
recording. Or, if a student's code passes all tests you can start up a
recording prompting them to go on to the next exercise.

## Current Limitations of Jupyter Graffiti

* Jupyter Graffiti can record most activities in Notebooks, but it does not record interactions with Jupyter's UX, e.g. you will not see the Jupyter menus get pulled down even if you pulled down a Jupyter menu during a recording.
* If you rearrange cells after making a recording, scrolling will try to align the cursor and the page as best it can with the cells you were mousing over and scrolling to, even if they are in a different order than when you made the original recording. However, due to complexities involving cell sizing, this process may not always be perfect.
* Copying cells does not copy their Graffiti.
* Make a Copy ... of the current Notebook will not create a copy of the recordings; in fact, it will use the same recording ID and therefore supplant recordings on the original Notebook. You can use an API call to fix this issue, however.
* Given this is the first version of this software, there may well be bugs. Feel free to report issues on Github and/or propose PR's.

## Future Plans

* In the next version of *Jupyter Graffiti* you will be able to
  automatically transcribe your spoken audio into subtitles that
  scroll along with the movie.

* Make a Copy ... of a Notebook should copy the recordings to a new Notebook recording ID.

### CREDITS

Author/Maintainer: 

Will Kessler 

github: <a href="https://github.com/willkessler">willkessler</a>
twitter: @atlas3650


Advisors: 

* Andy Brown (feature design)
* Tugce Akin (engineering)
* Nathan Tate (engineering).


RELEASES.md
### Version 1.0.2: Jan 18, 2020
* You can now hide the Graffiti Control panel menubar using notebook metadata (for Graffiti distribution)
* Fix to terminal function _Jump to Notebook's Dir_ so that it correctly goes to the right notebook directory
* Added directory help build for pip distribution.

### Version 1.0.1: Apr 29, 2019

#### New features:
* You can now specify the path to the Graffiti data directory and the name of the data directory itself, with a new icon available in the Graffiti Editor panel.

#### Other:
* `samples` directory and `user_manual` directory were cleaned up
* `images` were cleaned up

### Version 1.0.0: Apr 16, 2019

First fully functional open-sourced version of Jupyter Graffiti released for public use.



jupytergraffiti/js/svg.js
define([], () => {
  // Thanks to https://stackoverflow.com/questions/3642035/jquerys-append-not-working-with-svg-element
  const svg = {

    // Cf : https://www.beyondjava.net/how-to-connect-html-elements-with-an-arrow-using-svg
    generateArrowHeadElem: (color) => {
      const defs = document.createElementNS('http://www.w3.org/2000/svg', 'defs');
      const marker = svg.makeSvgElement('marker', {
        id:'arrowHead',
        viewBox: '0 0 10 10',
        refX: 0,
        refY: 5,
        markerUnits: 'strokeWidth',
        markerWidth: '10',
        markerHeight: '8',
        orient: 'auto',
        stroke: color
      });
      const path = svg.makeSvgElement('path', {
        d: "M 0 0 L 10 5 L 0 10 z"
      });
      marker.appendChild(path);
      defs.appendChild(marker);
      return defs;
    },
    
    /*
       const arrowHeadDef =
       '<defs>' +
       '  <marker id="arrowHead" viewBox="0 0 10 10" refX="0" refY="5" markerUnits="strokeWidth" markerWidth="10" markerHeight="8" orient="auto">' +
       '    <path d="M 0 0 L 10 5 L 0 10 z"></path>' +
       '  </marker>' +
       '</defs>';
       },
     */

    makeElementHtml: (tag, attr, innerHtml) => {
      let svgHtml = '<' + tag + ' ';
      if (tag === 'svg') {
        svgHtml += 'xmlns="http://www.w3.org/2000/svg" version="1.1" class="graffitiSvg" ';
      }
      let attrHtml = '';
      if (attr !== undefined) {
        attrHtml = $.map(attr, (val, key) => { return (key + '="' + val + '"') } ).join(' ');
      }
      if (innerHtml !== undefined) {
        svgHtml += attrHtml + '>' + innerHtml + '</' + tag + '>';
      } else {
        svgHtml += attrHtml + '></' + tag + '>';
      }
      return svgHtml;
    },

    makeSvgElement: (tag, attrs) => {
      const el= document.createElementNS('http://www.w3.org/2000/svg', tag);
      if (tag === 'svg') {
        el.setAttribute('xmlns', 'http://www.w3.org/2000/svg');
        el.setAttribute('version', '1.1');
        el.setAttribute('preserveAspectRatio', 'none')
      }
      for (let k in attrs) {
        el.setAttribute(k, attrs[k]);
      }
      return el;
    },

    makeDomElement: (tag, attrs) => {
      const el= document.createElement(tag);
      for (let k in attrs) {
        el.setAttribute(k, attrs[k]);
      }
      return el;
    },

    renderSvg: (svgChildren, x, y, width, height, viewBox) => {
      let containerDiv, containerSvg;
      let svgGenerator = $('#graffitiSvgGenerator');
      if (svgGenerator.length === 0) {
        $('body').append($('<div id="graffitiSvgGenerator"></div>'));
        svgGenerator = $('#graffitiSvgGenerator');
      }
      for (let svgChild of svgChildren) {
        let transform = '';
        if (svgChild.hasOwnProperty('cssTransform')) {
          transform = 'transform:' + svgChild.cssTransform;
        }
        containerDiv =
          svg.makeDomElement('div',
                             {
                               'class':"graffiti-svg-inner",
                               'style' : 'position:absolute;' +
                                         'left:' + svgChild.x + 'px;top:' + svgChild.y + 'px;' +
                                         'width:' + svgChild.width + 'px;height:' + svgChild.height + 'px;' +
                                         transform
                             });
        containerSvg =
          svg.makeSvgElement('svg',
                             {
                               width: svgChild.width,
                               height: svgChild.height,
                               viewBox: svgChild.viewBox
                             });
        if (svgChild.arrowAtEnd) {
          containerSvg.appendChild(svg.generateArrowHeadElem(svgChild.color));
        }
        containerSvg.appendChild(svgChild.el);
        containerDiv.appendChild(containerSvg);
        svgGenerator[0].appendChild(containerDiv);
      }
      const containerHtml = svgGenerator[0].innerHTML;
      svgGenerator.empty();

      return containerHtml;
    },

    makeLine: (opts) => {
      const viewBox = '0 0 1.12 1.12';
      const color = (opts.color === undefined ? '#000' : opts.color);
      const strokeWidth = (opts.strokeWidth === undefined ? 3 : opts.strokeWidth);
      let pathObj = 
        {
          'vector-effect': 'non-scaling-stroke',
          'stroke-width' : strokeWidth,
          stroke: color,
          fill: 'none',
          d: 'M ' + opts.p1x + ' ' + opts.p1y + ' L ' + opts.p2x + ' ' + opts.p2y
        };
      if (opts.arrowAtEnd) {
        pathObj['marker-end'] =  'url(#arrowHead)';
      }
      if (opts.dashed) {
        if (opts.dashWidth) {
          pathObj['stroke-dasharray'] = opts.dashWidth;
        } else {
          pathObj['stroke-dasharray'] = 4;
        }
      }
      const line = svg.makeSvgElement('path', pathObj);

      const renderedSvg = svg.renderSvg([
        {
          el: line,
          width: opts.width,
          height: opts.height,
          color: color,
          viewBox: viewBox,
          x: opts.x,
          y : opts.y,
          arrowAtEnd: true,
        }
      ]);
      return renderedSvg;

      
    },


    makeEllipse: (opts) => {
    },

    // need to use html injection, not code generation to make stuff work,
    // cf my post: https://stackoverflow.com/questions/52675823/preserveaspectratio-ignored-by-code-generation-but-not-html-injection-for-svg-p

    makeRightCurlyBracePath: () => {
      const rightCurlyBracePath =
        svg.makeSvgElement('path',
                           {
                             fill: "none",
                             stroke: "#000",
                             "vector-effect": "non-scaling-stroke",
                             "stroke-width" : "3",
                             d: "M0,0 A100, 173.20508075688772 0 0 1  100, 173.20508075688772 A100, " +
                                "173.20508075688772 0 0 0 200 346.41016151377545 A100, " +
                                "173.20508075688772 0 0 0 100, 519.6152422706632 A100, " +
                                "173.20508075688772 0 0 1 0, 692.8203230275509"
                           }
        );
      return rightCurlyBracePath;
    },

    makeLeftCurlyBrace: (x, y, height) => {
      const curlyViewBox = '0 0 200 692';
      const rightCurlyBracePath = svg.makeRightCurlyBracePath();
      const renderedSvg = svg.renderSvg([
        {
          el: rightCurlyBracePath,
          width: 8,
          height: height,
          viewBox: curlyViewBox,
          x: x,
          y : y,
          cssTransform: "scaleX(-1)" // css transform
        }
      ]);
      return renderedSvg;
    },

    makeRightCurlyBrace: (x, y, height) => {
      const curlyViewBox = '0 0 200 692';
      const rightCurlyBracePath = svg.makeRightCurlyBracePath();
      const renderedSvg = svg.renderSvg([
        {
          el: rightCurlyBracePath,
          width: 8,
          height: height,
          viewBox: curlyViewBox,
          x: x,
          y : y
        }
      ]);
      return renderedSvg;
    },

    makeSymmetricCurlyBraces: (x, y, width, height) => {
      const curlyViewBox = '0 0 200 692';
      const curlyBracePath1 = svg.makeRightCurlyBracePath();
      const curlyBracePath2 = svg.makeRightCurlyBracePath();
      const renderedSvg = svg.renderSvg([
        {
          el: curlyBracePath1,
          width: 8,
          height: height,
          viewBox: curlyViewBox,
          x: 0,
          y : 0,
          cssTransform: "scaleX(-1)" // css transform
        },
        {
          el: curlyBracePath2,
          width: 8,
          height: height,
          viewBox: curlyViewBox,
          x: width - 10,
          y : 0
        }
      ]);
      //console.log(renderedSvg);
      return renderedSvg;
    },

    // checkmark
    makeCheckmark: (x, y, width, height) => {
      const viewBox = '0 0 587 783.75';
      const checkmarkPath =
        svg.makeSvgElement('path',
                           {
                             fill: "#00aa00",
                             "stroke-width":"4",
                             d: "M0 303c61,65 122,129 184,194 134,-166 227,-376 403,-497 -181,160 -285,402 -400,627 -62,-108 -125,-216 -187,-324z"
                           }
        );

      const renderedSvg = svg.renderSvg([
        {
          el: checkmarkPath,
          width: width,
          height: height,
          viewBox: viewBox,
          x: x,
          y : y
        }
      ]);
      return renderedSvg;
    },

    // x (wrong) symbol
    makeX: (x, y, size) => {
      const viewBox = '0 0 233 291.25';
      const xPath =
        svg.makeSvgElement('polygon',
                           {
                             fill: "#aa0000",
                             points: "233,22 211,0 117,94 22,0 0,22 94,117 0,211 22,233 117,139 211,233 233,211 139,117"
                           }
        );

      const renderedSvg = svg.renderSvg([
        {
          el: xPath,
          width: size,
          height: size,
          viewBox: viewBox,
          x: x,
          y : y
        }
      ]);
      return renderedSvg;
    },

    makeSimplePath: (opts) => {
      const viewBox = opts.viewBox;
      const color = (opts.color === undefined ? '#000' : opts.color);
      const fill = (opts.fill === undefined ? 'none' : opts.fill);
      const strokeWidth = (opts.strokeWidth === undefined ? 3 : opts.strokeWidth);
      let pathObj = 
        {
          'vector-effect': 'non-scaling-stroke',
          'stroke-width' : strokeWidth,
          stroke: color,
          fill: fill,
          d: opts.d,
        };

      if (opts.dashWidth) {
        pathObj['stroke-dasharray'] = opts.dashWidth;
      } else {
        pathObj['stroke-dasharray'] = 4;
      }

      const thePath = svg.makeSvgElement('path',pathObj);

      const renderedSvg = svg.renderSvg([
        {
          el: thePath,
          width: opts.width,
          height: opts.height,
          viewBox: opts.viewBox,
          x: opts.x,
          y : opts.y
        }
      ]);
      return renderedSvg;
    },

    makeTopBracket: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          viewBox: '0 0 10 10',
          d: "M 0 10 L 0 0 L 10 0 L 10 10"
        })
      );
    },

    makeBottomBracket: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          viewBox: '0 0 10 10',
          d: "M 0 0 L 0 10 L 10 10 L 10 0"
        })
      );
    },

    makeLeftBracket: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          viewBox: '0 0 10 10',
          d: "M 10 10 L 0 10 L 0 0 L 10 0"
        })
      );
    },

    makeRightBracket: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          viewBox: '0 0 10 10',
          d: "M 0 0 L 10 0 L 10 10 L 0 10"
        })
      );
    },

    makeRectangle: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          viewBox: '0 0 100 100',
          d: "M 0 0 L 0 100 L 100 100 L 100 0 Z"
        })
      );
    },

    makeRightTriangle: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          viewBox: '0 0 100 125',
          d: "M 0 125 L 100 125 L 100 0 Z"
        })
      );
    },

    makeIsocelesTriangle: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          viewBox: '0 0 100 125',
          d: "M 0 125 L 100 125 L 50 0 Z",
        })
      );
    },

    makeTheta: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          viewBox: '0 0 469 843.75',
          d: "M469 334c-2,440 -453,468 -469,2 -13,-435 472,-460 469,-2zm-383 -20l298 0c-9,-366 -288,-376 -298,-6l0 6zm297 46l-297 0c16,345 279,397 297,11 0,-4 0,-7 0,-11z",
          fill: 'fill',
        })
      );
    },

    makeSigma: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          strokeWidth: 2,
          viewBox: '0 0 16 20',
          d: 'M2,1l5.46,7.27L2,15h12v-4h-1v1c0,0.552-0.448,1-1,1H4.97l4.39-5.52L5.25,2H12c0.552,0,1,0.448,1,1v1h1V1H2z',
          fill: 'fill',
        })
      );
    },

    makeSmiley: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          strokeWidth: 2,
          viewBox: '0 0 50 62.5',
          d: 'M25,1A24,24,0,1,0,49,25,24,24,0,0,0,25,1Zm0,46A22,22,0,1,1,47,25,22,22,0,0,1,25,47ZM35.77,33.32a1,1,0,0,1-.13,1.41C31.73,38,28.06,39.1,24.9,39.1a16,16,0,0,1-10.63-4.45,1,1,0,0,1,1.45-1.38c0.34,0.35,8.35,8.52,18.63-.08A1,1,0,0,1,35.77,33.32ZM15,19a3,3,0,1,1,3,3A3,3,0,0,1,15,19Zm14,0a3,3,0,1,1,3,3A3,3,0,0,1,29,19Z',
          fill: 'none',
        })
      );
    },

    makeFrowney: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          strokeWidth: 2,
          viewBox: '0 0 100 125',
          d: 'M50,2.5C23.809,2.5,2.5,23.809,2.5,50S23.809,97.5,50,97.5S97.5,76.191,97.5,50S76.191,2.5,50,2.5z M50,91.684    C27.016,91.684,8.316,72.984,8.316,50S27.016,8.316,50,8.316S91.684,27.016,91.684,50S72.984,91.684,50,91.684z M37.489,41.386    c2.964,0,5.369-2.403,5.369-5.369c0-2.966-2.405-5.368-5.369-5.368c-2.966,0-5.369,2.402-5.369,5.368    C32.12,38.982,34.523,41.386,37.489,41.386z M62.511,41.386c2.965,0,5.369-2.403,5.369-5.369c0-2.966-2.404-5.368-5.369-5.368    c-2.966,0-5.368,2.402-5.368,5.368C57.143,38.982,59.545,41.386,62.511,41.386z M50.001,51.186    c-13.939,0-20.525,9.548-22.06,14.597c-0.467,1.537,0.399,3.161,1.936,3.628c1.539,0.471,3.161-0.399,3.628-1.936    c0.032-0.105,3.336-10.473,16.496-10.473c13.015,0,16.363,10.061,16.494,10.472c0.381,1.255,1.534,2.063,2.781,2.063    c0.28,0,0.564-0.04,0.846-0.127c1.538-0.467,2.405-2.091,1.938-3.627C70.524,60.733,63.939,51.186,50.001,51.186z',
          fill: 'none',
        })
      );
    },

    makeThumbsUp: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          strokeWidth: 2,
          viewBox: '0 0 218 346.25',
          d: 'M28 263l31 -9c64,42 77,13 92,10 4,0 1,4 17,0 22,-7 31,-19 23,-35 19,-6 21,-18 15,-33 15,-9 15,-26 3,-38 19,-37 -11,-67 -80,-48 -5,-36 11,-59 5,-80 -7,-27 -25,-31 -50,-30 3,68 8,35 -25,101 -27,55 -3,48 -57,63 -6,36 4,70 26,99zm4 -12c-16,-24 -23,-49 -21,-77 48,-14 33,-15 57,-65 33,-71 31,-34 27,-97 31,1 32,26 26,50 -7,27 -6,40 -1,62 26,-7 74,-21 82,6 7,27 -22,40 -35,41l-42 -7c9,-28 36,-19 44,-19l10 -3 7 -13c-29,8 -10,3 -31,4 -24,1 -40,15 -43,40l8 1c-8,7 -13,16 -14,28l9 1c-5,6 -10,15 -12,26l14 3c-5,7 -9,15 -11,26l29 4c-29,10 -50,-1 -74,-20l-29 9zm87 -58c12,-30 27,-10 49,-12 5,0 27,-7 33,-14 24,20 -36,32 -39,33l-43 -7zm-2 27l10 -15c44,7 28,8 70,-4 10,19 -35,26 -35,26l-45 -7zm3 30l9 -17c36,5 26,7 53,0 4,16 -17,22 -23,22l-39 -5z',
          fill: 'fill',
        })
      );
    },

    makeThumbsDown: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          strokeWidth: 2,
          viewBox: '0 0 226 357.5',
          d: 'M18 27l33 4c59,-51 77,-23 92,-23 5,0 0,-4 18,-2 23,3 33,15 28,32 20,4 24,15 20,32 17,7 19,23 8,37 25,36 -1,70 -74,60 0,38 19,59 16,82 -3,27 -21,34 -46,37 -6,-70 3,-37 -40,-99 -35,-52 -10,-48 -67,-56 -11,-36 -6,-71 12,-104zm6 12c-13,26 -16,53 -10,81 51,7 35,11 67,58 44,66 36,29 41,95 32,-7 29,-32 19,-55 -11,-27 -11,-40 -9,-63 27,4 78,10 82,-18 4,-28 -28,-37 -41,-36l-42 13c13,27 39,14 48,13l10 1 9 12c-30,-4 -10,-1 -33,1 -23,2 -41,-9 -49,-34l8 -3c-8,-6 -15,-14 -18,-26l9 -3c-6,-5 -11,-13 -16,-24l15 -5c-7,-6 -12,-15 -15,-25l28 -8c-30,-7 -50,8 -72,30l-31 -4zm96 46c17,28 30,7 52,5 6,-1 29,4 36,9 21,-23 -42,-27 -44,-27l-44 13zm-5 -27l12 14c44,-13 27,-12 72,-6 7,-21 -39,-22 -40,-22l-44 14zm-1 -31l11 16c36,-11 26,-11 54,-7 2,-18 -20,-20 -27,-20l-38 11z',
          fill: 'fill',
        })
      );
    },


    makeStar: (opts) => {
      return svg.makeSimplePath(
        $.extend(opts, {
          strokeWidth: 2,
          viewBox: '0 0 1024 1280',
          d:
            'M521.19122173 257.59668148l48.67463902 112.36592198c10.61521383 24.46677333 33.65799506 41.42522469 60.32548344 44.14375505l123.62840495 12.55702124-92.30057877 79.48464988c-20.71261235 17.86462815-29.90383408 45.43829333-23.8195042 72.10578173l27.44421136 121.68659752-105.37541531-64.20909827c-11.52139061-6.99050667-24.85513482-10.74466765-38.31833283-10.74466765-13.59265185 0-26.79694222 3.75416098-38.31833286 10.74466765l-105.37541529 64.20909827 27.44421135-121.68659752c5.95487605-26.66748839-3.10689185-54.24115358-23.8195042-72.10578173l-92.30057876-79.48464988 123.62840494-12.55702124c26.53803457-2.71853037 49.71026963-19.5475279 60.32548346-44.14375505l48.1568237-112.36592198m0-117.80298272c-6.2137837 0-12.55702124 3.3657995-15.40500543 10.09739852l-85.43952593 197.28763258c-2.45962272 5.56651457-7.63777581 9.45012939-13.72210568 10.09739853l-216.446799 22.00715063c-14.7577363 1.55344592-20.45370469 19.80643555-9.32067556 29.51547258l163.11182222 140.45740248c4.66033778 4.01306864 6.86105283 10.35630617 5.43706074 16.44063605l-48.1568237 213.98717629c-2.58907653 11.26248297 6.34323753 20.58315852 16.44063604 20.58315852 2.84798419 0 5.95487605-0.77672297 8.67340642-2.45962272l186.15460346-113.40155259c2.71853037-1.68289975 5.69596839-2.45962272 8.80286024-2.45962272s6.08432987 0.77672297 8.80286026 2.45962272l186.15460344 113.40155259c2.84798419 1.68289975 5.82542222 2.45962272 8.67340644 2.45962272 10.09739852 0 19.02971259-9.32067555 16.44063604-20.58315852L693.23535803 565.69679013c-1.4239921-6.08432987 0.77672297-12.42756741 5.43706073-16.44063605l163.11182222-140.45740248c11.26248297-9.70903703 5.43706075-27.96202667-9.32067555-29.51547258l-216.44679901-22.00715063c-6.08432987-0.64726914-11.26248297-4.40143013-13.72210567-10.09739853l-85.43952593-197.28763258c-3.23634569-6.73159902-9.45012939-10.09739852-15.66391309-10.09739852z',
          fill: 'fill',
        })
      );
    },

  }

  return (svg);

});





